packages/query-core/src/types.ts

```diff
@@ -841,3 +843,29 @@ export type NotifyEventType =
 export interface NotifyEvent {
   type: NotifyEventType
 }
+
+export type MutationActionType =
+  | 'failed'
+  | 'pending'
+  | 'success'
+  | 'error'
+  | 'pause'
+  | 'continue'
+
+export type QueryActionType =
+  | 'failed'
+  | 'fetch'
+  | 'success'
+  | 'error'
+  | 'invalidate'
+  | 'pause'
+  | 'continue'
+  | 'setState'
```

Comment:
the idea is that `Action` defines the potential actions, and the type can be inferred by doing:

```
type QueryAction = Action["type"]
```

this makes `Action` the single source of truth. If we add one action, it also becomes a valid type. With this change, we'd need to define it in two places. I don't see the advantage of that.

Reply:
```ts
type QueryActionType<TData, TError> = Action<TData, TError>['type']
```
Does this mean that it can be utilized in the same way as above? I understood your comment.

https://github.com/TanStack/query/pull/6309#discussion_r1382442647 
I initially thought it was safer because it throws a type error when written incompatibly.

Crucially, I also thought it was the right thing to do for consistency, since "queryCache" was also done this way.

```ts
interface NotifyEventQueryAdded extends NotifyEvent {
  type: 'added'
  query: Query<any, any, any, any>
}
```
```ts
export type NotifyEventType =
  | 'added'
  | 'removed'
  | 'updated'
  | 'observerAdded'
  | 'observerRemoved'
  | 'observerResultsUpdated'
  | 'observerOptionsUpdated'

export interface NotifyEvent {
  type: NotifyEventType
}
```

or `QueryObserverBaseResult`

```ts
export type QueryStatus = 'pending' | 'error' | 'success'
export type FetchStatus = 'fetching' | 'paused' | 'idle'

export interface QueryObserverBaseResult<
  TData = unknown,
  TError = DefaultError,
> {
  // ...
  status: QueryStatus
  fetchStatus: FetchStatus
}
```
```ts
export interface QueryObserverLoadingResult<
  TData = unknown,
  TError = DefaultError,
> extends QueryObserverBaseResult<TData, TError> {
  // ...
  status: 'pending'
}

export interface QueryObserverLoadingErrorResult<
  TData = unknown,
  TError = DefaultError,
> extends QueryObserverBaseResult<TData, TError> {
  // ...
  status: 'error'
}
```

Reply:
@TkDodo Doesn't it make sense to make our code consistent?

Reply:
those are public interfaces, and we need the event types for queries and mutations to be the same